File: cpp/im/forward_models.cpp
// This source code is licensed under the 3-clause BSD license found
// in the LICENSE file in the root directory of this project.

#include "forward_models.h"

#include <common/b_spline_2d.h>
#include <common/binning_table.h>
#include <common/ckd.h>
#include <common/fourier.h>
#include <common/isrf.h>
#include <common/l1.h>
#include <netcdf>
#include <random>
#include <spdlog/spdlog.h>

namespace tango {

auto applyISRF(const CKD& ckd,
               const bool enabled,
               const ISRF& isrf,
               const std::string& sgm_filename,
               L1& l1_prod) -> void
{
    l1_prod.level = ProcLevel::l1b;
    // If this process is disabled then linearly interpolate the
    // line-by-line spectra onto the CKD wavelength grids. We cannot
    // simply return like the other processes.
    const size_t n_waves_in { static_cast<size_t>(l1_prod.wavelengths.size()) };
    const size_t n_waves_out { static_cast<size_t>(
      ckd.wave.wavelengths.size()) };
    ArrayXXd spectra_out(l1_prod.n_alt * ckd.n_act, n_waves_out);
    if (enabled) {
        // If the spectra are not in memory then read them one by one
        // for the convolution.
        if (l1_prod.spectra.size() == 0) {
            const netCDF::NcFile nc { sgm_filename, netCDF::NcFile::read };
            const netCDF::NcVar nc_var { nc.getVar("radiance") };
            ArrayXXd buf(ckd.n_act, n_waves_in);
            for (int i_alt {}; i_alt < l1_prod.n_alt; ++i_alt) {
                nc_var.getVar({ static_cast<size_t>(i_alt), 0, 0 },
                              { 1, static_cast<size_t>(ckd.n_act), n_waves_in },
                              buf.data());
#pragma omp parallel for
                for (int i_act = 0; i_act < ckd.n_act; ++i_act) {
                    const int act_idx { i_alt * ckd.n_act + i_act };
                    spectra_out.row(act_idx) =
                      isrf.convolve(buf.row(i_act), act_idx);
                }
            }
        } else {
#pragma omp parallel for
            for (int i = 0; i < l1_prod.n_alt * ckd.n_act; ++i) {
                spectra_out.row(i) = isrf.convolve(l1_prod.spectra.row(i), i);
            }
        }
    } else {
#pragma omp parallel for
        for (int i_act = 0; i_act < l1_prod.n_alt * ckd.n_act; ++i_act) {
            const LinearSpline spline { l1_prod.wavelengths,
                                        l1_prod.spectra.row(i_act) };
            for (size_t i_wave {}; i_wave < n_waves_out; ++i_wave) {
                spectra_out.row(i_act)(i_wave) =
                  spline.eval(ckd.wave.wavelengths(i_wave));
            }
        }
    }
    l1_prod.spectra = std::move(spectra_out);
    l1_prod.wavelengths = ckd.wave.wavelengths;
}

auto radiometric(const CKD& ckd, const bool enabled, L1& l1_prod) -> void
{
    l1_prod.level = ProcLevel::swath;
    if (!enabled) {
        return;
    }
    const double exposure_time_inv { 1.0 / l1_prod.exposure_time };
#pragma omp parallel for
    for (int i_alt = 0; i_alt < l1_prod.n_alt; ++i_alt) {
        l1_prod.spectra(Eigen::seqN(i_alt * ckd.n_act, ckd.n_act),
                        Eigen::all) /= ckd.rad.rad * exposure_time_inv;
    }
}

auto mapToDetector(const CKD& ckd,
                   const int b_spline_order,
                   L1& l1_prod) -> void
{
    l1_prod.level = ProcLevel::stray;
    l1_prod.signal.resize(l1_prod.n_alt, ckd.npix);
#pragma omp parallel for
    for (int i_alt = 0; i_alt < l1_prod.n_alt; ++i_alt) {
        const BSpline2D bspline_2d {
            b_spline_order,
            ckd.swath.act_angles,
            ckd.wave.wavelengths,
            l1_prod.spectra(Eigen::seqN(i_alt * ckd.n_act, ckd.n_act),
                            Eigen::all)
        };
        bspline_2d.eval(ckd.swath.act_map,
                        ckd.swath.wavelength_map,
                        l1_prod.signal.row(i_alt));
        for (int i {}; i < ckd.npix; ++i) {
            l1_prod.signal(i_alt, i) = std::max(0.0, l1_prod.signal(i_alt, i));
        }
    }
    l1_prod.spectra = ArrayXXd {};
}

auto strayLight(const CKD& ckd, const bool enabled, L1& l1_prod) -> void
{
    l1_prod.level = ProcLevel::prnu;
    if (!enabled) {
        return;
    }
    // Pre-allocate the signal FFT for efficiency
    Eigen::ArrayXcd signal_fft(ckd.stray.kernel_fft_sizes.maxCoeff());
    // Result of convolution
    ArrayXXd conv_result(ckd.n_detector_rows, ckd.n_detector_cols);
#pragma omp parallel for firstprivate(signal_fft, conv_result)
    for (int i_alt = 0; i_alt < l1_prod.n_alt; ++i_alt) {
        auto signal = l1_prod.signal.row(i_alt).reshaped<Eigen::RowMajor>(
          ckd.n_detector_rows, ckd.n_detector_cols);
        if (ckd.stray.n_kernels > 1) {
            convolveMulti(ckd.stray.edges,
                          ckd.stray.weights,
                          signal,
                          ckd.stray.kernel_rows,
                          ckd.stray.kernel_cols,
                          ckd.stray.kernels_fft,
                          signal_fft,
                          conv_result);
        } else {
            convolve(signal,
                     ckd.stray.kernel_rows(0),
                     ckd.stray.kernel_cols(0),
                     ckd.stray.kernels_fft.front(),
                     signal_fft,
                     conv_result);
        }
        signal = (1.0 - ckd.stray.eta) * signal + conv_result;
        l1_prod.signal.row(i_alt) = signal.reshaped<Eigen::RowMajor>();
    }
}

auto prnu(const CKD& ckd, const bool enabled, L1& l1_prod) -> void
{
    l1_prod.level = ProcLevel::nonlin;
    if (!enabled) {
        return;
    }
#pragma omp parallel for
    for (int i_alt = 0; i_alt < l1_prod.n_alt; ++i_alt) {
        l1_prod.signal.row(i_alt) *= ckd.prnu.prnu;
    }
}

auto nonlinearity(const bool enabled,
                  const LinearSpline& nonlin_spline,
                  L1& l1_prod) -> void
{
    l1_prod.level = ProcLevel::dark_current;
    if (!enabled) {
        return;
    }
#pragma omp parallel for
    for (int i_alt = 0; i_alt < l1_prod.n_alt; ++i_alt) {
        l1_prod.signal.row(i_alt) =
          nonlin_spline.eval(l1_prod.signal.row(i_alt));
    }
}

auto darkCurrent(const CKD& ckd, const bool enabled, L1& l1_prod) -> void
{
    l1_prod.level = ProcLevel::noise;
    if (!enabled) {
        return;
    }
#pragma omp parallel for
    for (int i_alt = 0; i_alt < l1_prod.n_alt; ++i_alt) {
        l1_prod.signal.row(i_alt) += ckd.dark.current * l1_prod.exposure_time;
    }
}

auto noise(const CKD& ckd,
           const bool enabled,
           const int seed,
           const int n_coadditions,
           L1& l1_prod) -> void
{
    l1_prod.level = ProcLevel::dark_offset;
    if (!enabled) {
        return;
    }
#pragma omp parallel for
    for (int i_alt = 0; i_alt < l1_prod.n_alt; ++i_alt) {
        static std::mt19937 gen { static_cast<unsigned long>(seed) };
        for (int i {}; i < ckd.npix; ++i) {
            const double noise_value { std::sqrt(
              (ckd.noise.n2(i)
               + std::abs(l1_prod.signal(i_alt, i)) * ckd.noise.g(i))
              / n_coadditions) };
            std::normal_distribution<> d { 0.0, noise_value };
            l1_prod.signal(i_alt, i) += d(gen);
        }
    }
}

auto darkOffset(const CKD& ckd, const bool enabled, L1& l1_prod) -> void
{
    l1_prod.level = ProcLevel::raw;
    if (!enabled) {
        return;
    }
#pragma omp parallel for
    for (int i_alt = 0; i_alt < l1_prod.n_alt; ++i_alt) {
        l1_prod.signal.row(i_alt) += ckd.dark.offset;
    }
}

auto binDetectorImages(const int n_rows,
                       const int n_cols,
                       const std::string& binning_filename,
                       const int binning_table_id,
                       const bool scale_by_binsize,
                       L1& l1_prod) -> void
{
    const BinningTable binning_table {
        n_rows, n_cols, binning_filename, binning_table_id
    };
    l1_prod.signal = binning_table.binMulti(l1_prod.signal, scale_by_binsize);
    l1_prod.binning_table_id = static_cast<uint8_t>(binning_table_id);
}

auto digitalToAnalog(const int nr_coadditions, L1& l1_prod) -> void
{
    l1_prod.level = ProcLevel::l1a;
    l1_prod.nr_coadditions = static_cast<uint16_t>(nr_coadditions);
    l1_prod.signal *= l1_prod.nr_coadditions;
    l1_prod.signal = Eigen::round(l1_prod.signal);
}

auto estimateOptimalCoadd(const CKD& ckd,
                          const int FMC,
                          const double exposure_time,
                          const double f_sat,
                          const double full_well,
                          const double t_dead,
                          const L1& l1_prod) -> void
{
    constexpr double dwell { 24.0 };
    const double t_dwell { FMC / dwell };
    for (int i_alt = 0; i_alt < l1_prod.n_alt; ++i_alt) {
        double max_val {};
        for (int i {}; i < ckd.npix; ++i) {
            if (!ckd.pixel_mask(i)) {
                max_val = std::max(max_val, l1_prod.signal(i_alt, i));
            }
        }
        const double I_sig { max_val / exposure_time };
        const double coadd_raw { I_sig * t_dwell
                                 / (f_sat * full_well + I_sig * t_dead) };
        const int n_coadd { static_cast<int>(std::ceil(coadd_raw)) };
        spdlog::info("  {:5.2f} ({}) {:6.4} ms",
                     coadd_raw,
                     n_coadd,
                     1e3 * (t_dwell / n_coadd) - t_dead);
    }
}

} // namespace tango

File: teds/im/forward_models.py
# This source code is licensed under the 3-clause BSD license found in
# the LICENSE file in the root directory of this project.
"""List of processes applied to data by the instrument model.

Each process undoes or "uncalibrates" a step that is performed by the
L1A-L1B processor, gradually bringing the data level from L1B to L1A
or anywhere in between.

"""
from netCDF4 import Dataset
from scipy.interpolate import interpn
from tqdm import tqdm
import numpy as np
import numpy.typing as npt

from .types import Stokes
from teds.l1al1b.calibration import convolve_with_all_kernels
from teds.l1al1b.types import BinningTable
from teds.l1al1b.types import CKDNoise
from teds.l1al1b.types import CKDNonlin
from teds.l1al1b.types import CKDStray
from teds.l1al1b.types import CKDSwath
from teds.l1al1b.types import L1
from teds.l1al1b.types import ProcLevel
from teds.lib.convolution import Kernel


def apply_mueller(l1_product: L1,
                  stokes: Stokes,
                  mueller: npt.NDArray[np.floating]) -> None:
    """Add Mueller matrix elements to Stokes I,Q,U parameters.

    Compute I_pol = I + M_01 / M_00 * Q + M_02 / M_00 * U.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings)
    stokes
        Stokes Q and U parameters
    mueller
        Mueller matrix elements for each ALT, ACT, and wavelength

    """
    M_00 = mueller[0]
    M_01 = mueller[1]
    M_02 = mueller[2]
    l1_product.spectra += M_01 / M_00 * stokes.Q + M_02 / M_00 * stokes.U


def apply_isrf(l1_product: L1,
               isrf: Kernel,
               convolve: bool,
               sgm_filename: str,
               alt_beg: int) -> None:
    """Convolve spectra with the ISRF.

    The ISRF has a fixed shape as a function of wavelength and does not depend
    on pixel.

    For debugging or experimentation, the ISRF convolution can be
    turned off in which case this function interpolates onto a target
    grid without convolving.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    isrf
        Convolution kernel
    convolve
        Convolve with ISRF (True) or only interpolate (False)
    sgm_filename
        SGM radiance filename. Only required if the spectra are not in
        memory. Then read them one by one for ISRF convolution.
    alt_begin
        First ALT position for convolution if spectra are read from
        the SGM file. The last position can be derived from the
        dimensions of the allocated spectra variable.

    """
    l1_product.proc_level = ProcLevel.l1b
    n_alt = l1_product.spectra.shape[0]
    n_act = l1_product.spectra.shape[1]
    conv = np.empty((n_alt, n_act, len(isrf.wavelengths_out)))
    if convolve:
        # If the spectra are not in memory then read them one by one
        # for the convolution.
        in_memory = l1_product.spectra.shape[-1] > 0
        if not in_memory:
            nc = Dataset(sgm_filename)
        for i_alt in tqdm(range(n_alt), unit=' ALT'):
            for i_act in range(n_act):
                if in_memory:
                    conv[i_alt, i_act, :] = isrf.convolve(
                        l1_product.spectra[i_alt, i_act, :])
                else:
                    conv[i_alt, i_act, :] = isrf.convolve(
                        nc['radiance'][alt_beg + i_alt, i_act, :].data)
        l1_product.spectra = conv
    else:
        for i_alt in tqdm(range(n_alt), unit=' ALT'):
            for i_act in range(n_act):
                conv[i_alt, i_act, :] = np.interp(
                    isrf.wavelengths_out,
                    l1_product.wavelengths,
                    l1_product.spectra[i_alt, i_act, :])
    l1_product.spectra = conv
    l1_product.wavelengths = isrf.wavelengths_out


def radiometric(l1_product: L1, rad_corr: npt.NDArray[np.floating]) -> None:
    """Convert from spectral photon radiance [nm-1 s-1 sr-1 m-2] to
    counts.

    The quantum efficiency [e/ph] is taken into account by the PRNU
    step.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    rad_corr
        Radiance responsivity correction factor [nm-1 sr-1 m-2].

    """
    l1_product.proc_level = ProcLevel.swath
    for i_alt in tqdm(range(l1_product.spectra.shape[0])):
        l1_product.spectra[i_alt, :, :] *= (
            l1_product.exposure_time / (rad_corr[0, 0]))


def map_to_detector(l1_product: L1,
                    ckd: CKDSwath,
                    wavelengths: npt.NDArray[np.floating],
                    b_spline_order: int) -> None:
    """Map spectra to detector.

    Spectra are mapped to infinitely thin curves on the detector (instead of
    simply to rows due to spatial distortion or keystone) and then interpolated
    along columns (not along constant wavelengths) at integer row coordinates,
    using cubic interpolation and linear extrapolation.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    ckd
        Swath section of the CKD
    wavelengths
        Main CKD wavelength grids (not the intermediate grids). Only
        used if exact_drawing is true.
    b_spline_order
        2D B-spline order

    """
    l1_product.proc_level = ProcLevel.stray
    n_alt = l1_product.spectra.shape[0]
    n_rows, n_cols = ckd.act_map.shape
    l1_product.signal = np.empty((n_alt, n_rows * n_cols))
    act_wavelength_map = np.column_stack((ckd.act_map.ravel(),
                                          ckd.wavelength_map.ravel()))
    b_spline_methods = {1: 'linear', 3: 'cubic', 5: 'quintic'}
    for i_alt in tqdm(range(n_alt)):
        l1_product.signal[i_alt, :] = interpn(
            (ckd.act_angles, wavelengths),
            l1_product.spectra[i_alt, :, :],
            act_wavelength_map,
            method=b_spline_methods[b_spline_order],
            bounds_error=False,
            fill_value=None).reshape(n_rows * n_cols)


def stray_light(l1_product: L1, ckd: CKDStray) -> None:
    """Add stray light to the signal.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    ckd
        Stray light CKD containing a list of the Fourier transforms of
        kernels, weights of subsignals, and an 'edges' array which
        specifies the location of each subsignal within the original
        signal.

    """
    l1_product.proc_level = ProcLevel.prnu
    n_alt = l1_product.signal.shape[0]
    eta = ckd.eta.ravel()
    for i_alt in tqdm(range(n_alt)):
        signal = l1_product.signal[i_alt]
        stray_convolved = convolve_with_all_kernels(signal, ckd)
        signal_convolved = (1 - eta) * signal + stray_convolved
        l1_product.signal[i_alt, :] = signal_convolved


def prnu(l1_product: L1, prnu_qe: npt.NDArray[np.floating]) -> None:
    """Incorporate PRNU and quantum efficiency.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    prnu_qe
        Detector map of PRNU times quantum efficiency (not the
        correction).

    """
    l1_product.proc_level = ProcLevel.nonlin
    l1_product.signal *= prnu_qe


def nonlinearity(l1_product: L1, ckd: CKDNonlin) -> None:
    """Incorporate non-linearity.

    Only a dependence on signal is taken into account, not on
    intensity, so make sure the non-linearity CKD is valid for the
    used exposure time.

    The model is a linear interpolation of a set of points, so make
    sure the number of points is large enough to produce a smooth
    curve. Input data outside the model range are clipped. The model
    does not depend on pixel.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    ckd
        Nonlinearity CKD consisting of an expected, linear signal
        [counts], and an observed, non-linear signal [counts].

    """
    l1_product.proc_level = ProcLevel.dark_current
    l1_product.signal = np.interp(
        l1_product.signal, ckd.expected, ckd.observed)


def dark_current(l1_product: L1,
                 dark_current: npt.NDArray[np.floating]) -> None:
    """Incorporate dark signal.

    If the dark signal does not depend linearly on exposure time, make
    sure the dark current CKD is valid for the used exposure time.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    dark_current
        Detector map of dark current [counts/s].

    """
    l1_product.proc_level = ProcLevel.noise
    l1_product.signal += dark_current * l1_product.exposure_time


def noise(l1_product: L1,
          ckd: CKDNoise,
          dark_current: npt.NDArray[np.floating],
          n_coadditions: int,
          seed: int) -> None:
    """Add random noise to signal.

    The signal is changed.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    ckd
        Noise CKD consisting of maps of read_noise [counts] and the
        conversion gain [e/counts].
    dark_current
        Detector map of dark current [counts/s].
    seed
        Seed for random number generator.

    """
    l1_product.proc_level = ProcLevel.dark_offset
    # The absolute value of dark_signal should be taken because a
    # negative signal still increases the noise.
    variance = ((ckd.read_noise**2 + l1_product.signal / ckd.conversion_gain)
                / n_coadditions)
    std = np.sqrt(np.clip(variance, 0, None))
    rng = np.random.default_rng(seed)
    l1_product.signal += rng.normal(0.0, std, l1_product.signal.shape)


def dark_offset(l1_product: L1, offset: npt.NDArray[np.floating]) -> None:
    """Incorporate offset.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    offset
        Detector map of offset [counts].

    """
    l1_product.proc_level = ProcLevel.raw
    l1_product.signal += offset


def bin_detector_images(l1_product: L1,
                        binning_table_id: int,
                        binning_table: BinningTable) -> None:
    """Bin all detector images.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    binning_table_id
        Binning factor
    binning_table
        Bin index of each pixel in an unbinned frame and number of
        pixels in each bin of a binned frame.

    """
    n_alt = l1_product.signal.shape[0]
    binned_signals = np.zeros((n_alt, len(binning_table.count_table)))
    for i_alt in range(l1_product.signal.shape[0]):
        for idx, idx_binned in enumerate(binning_table.bin_indices.ravel()):
            binned_signals[i_alt, idx_binned] += l1_product.signal[i_alt, idx]
    l1_product.signal = binned_signals
    l1_product.binning_table_id = binning_table_id


def coadd_and_adc(l1_product: L1, n_coadditions: int) -> None:
    """Coadd and convert detector images to integer format.

    Coaddition effectively makes exact copies of existing frames and
    sums them, i.e. if noise was added, that is exactly the same in
    each frame copy.

    Parameters
    ----------
    l1_product
        L1 product (signal and detector settings).
    n_coadditions
        Coaddition factor.

    """
    l1_product.proc_level = ProcLevel.l1a
    l1_product.coad_factor = n_coadditions
    l1_product.signal = l1_product.signal * l1_product.coad_factor
    # In reality, the noise realization is different in every read-out
    # and the signal is rounded before summing. Here the noise is not
    # calculated for each read-out separately. Then the best
    # approximation is to round after summing.
    l1_product.signal = np.round(l1_product.signal)
